guide: Persisting experiments (#2845)
iterative/dvc.org: content/docs/user-guide/experiment-management/persisting-experiments.md

# Persisting Experiments

DVC runs experiments outside of the standard Git stage/commit workflow for quick
iteration. In this section we'll see how to bring the experiments we run to the

> standard Git workflow with `dvc exp branch` and `dvc exp apply`.

Reviewer: Good, thanks! Just one thing: It's a bit repetitive to use the "std git flow" phrase in both sentences. Do we even need to focus so much on Git? Maybe there's a more general way to state this in the first sentence (then it can be OK to be specific about git flow in the 2nd one).

## Create a Git branch for your experiment

> You can use `dvc exp branch` to create a new branch from the experiment and keep
> all code and artifacts separate from your current one.

Reviewer: Copy edit
```suggestion
You can use `dvc exp branch` to create a new branch from an experiment, and keep
all its code and artifacts separate from your current <abbr>workspace</abbr>.
```
Keeping in mind there may be unstaged changes in the working tree.

> all code and artifacts separate from your current one.

Reviewer: Yeah I'd definitely keep the `apply` section first and `branch` after.

Reviewer: @dberenbaum so you recommend `exp branch` be introduced first as we want to encourage it more? Thanks

```dvc
$ dvc exp show --include-params=my_param
```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━┓
┃ neutral:**Experiment**            ┃ neutral:**Created**      ┃     metric:**auc** ┃ param:**my_param**   ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━┩
│ workspace             │ -            │ 0.61314 │ 3          │
│ experiments-tag       │ Oct 19, 2020 │ 0.61314 │ 3          │
│ ├── exp-e6c97         │ Oct 20, 2020 │ 0.69830 │ 2          │
│ └── exp-1df77         │ Oct 22, 2020 │ 0.51676 │ 1          │
└───────────────────────┴──────────────┴─────────┴────────────┘
```

Suppose you want to continue to work on `exp-e6c97` in a separate branch. You
can create a new Git branch by specifying the experiment and giving a new name
for it:

```dvc
$ dvc exp branch exp-e6c97 my-branch
Git branch 'my-branch' has been created from experiment 'exp-e6c97'.
To switch to the new branch run:
        git checkout my-branch
```

> Note that DVC doesn't checkout the new branch. You can one or more branches from
> existing experiments, and switch into any one manually like this:

Reviewer: Copy edit
```suggestion
Note that DVC doesn't switch into the new branch. You can create one or more branches from
existing experiments, and switch into any one manually like this:
```

```dvc
$ git checkout my-branch
$ dvc checkout
```

Your workspace now contains all the files from the experiment.

> ## Bring Experiment results to your workspace

Reviewer: CE
```suggestion
## Bring experiment results to your workspace
```

Typically, `dvc exp run` leaves the experiment results in your workspace for
convenience. However, you may have run multiple experiments and wish to go back
to a specific one. In this case, you can restore a previous experiment's results
with `dvc exp apply`. Let's see an example:

```dvc
$ dvc exp show --include-params=my_param
```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━┓
┃ neutral:**Experiment**            ┃ neutral:**Created**      ┃     metric:**auc** ┃ param:**my_param**   ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━┩
│ workspace             │ -            │ 0.61314 │ 3          │
│ experiments-tag       │ Oct 19, 2020 │ 0.61314 │ 3          │
│ ├── exp-e6c97         │ Oct 20, 2020 │ 0.69830 │ 2          │
│ └── exp-1df77         │ Oct 22, 2020 │ 0.51676 │ 1          │g
└───────────────────────┴──────────────┴─────────┴────────────┘
```

> The results found in the workspace can be found in the respective row. When you
> want to bring another experiment to the workspace, you can refer to it using

Reviewer: CE
```suggestion
The results found in the workspace are shown in the respective row. When you
want to bring another experiment to the workspace, you can reference it using
```

it's name or ID, e.g.:

```dvc
$ dvc exp apply exp-e6c97
Changes for experiment 'exp-e6c97' have been applied...

> ```

Reviewer: We may want to note that the working tree has to be clean (stash or commit changes).

Reviewer: > > the working tree has to be clean
>
> there is no such restriction

OK maybe not a hard restriction (not always), but `exp apply` will fail if there are conflicts between the experiment results and a dirty working tree, right?

Now, if you list the experiments again with `dvc exp show`, you'll see that the
workspace contains the experiment with `my_param` of `2` and corresponding `auc`
value.

You can now use standard Git commands (e.g. `git add/commit/push`) to version
this experiment directly in the <abbr>repository</abbr>. DVC-tracked data and
artifacts are already in the DVC cache, and the rest (params, code and config

> files, etc.) can be stored in Git. Please note that you may need to run
> `dvc push` to copy the DVC cache contents to your DVC remote to share or backup.

Reviewer: CE
```suggestion
files, etc.) can be stored in Git.

> Please note that you need to
`dvc push` in order to share or backup the DVC cache contents.
```

this experiment directly in the <abbr>repository</abbr>. DVC-tracked data and
artifacts are already in the DVC cache, and the rest (params, code and config
files, etc.) can be stored in Git.

## Create a Git branch for your experiment

You may desire to keep the experiments in their separate Git branches. You can
use `dvc exp branch` to create a new branch from the experiment and keep all
code and artifacts separate from your current one.

```dvc
$ dvc exp show --include-params=my_param
```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Experiment            ┃ Created      ┃     auc ┃ my_param   ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━┩
│ workspace             │ -            │ 0.61314 │ 3          │
│ experiments-tag       │ Oct 19, 2020 │ 0.61314 │ 3          │
│ ├── exp-e6c97         │ Oct 20, 2020 │ 0.69830 │ 2          │
│ └── exp-1df77         │ Oct 22, 2020 │ 0.51676 │ 1          │
└───────────────────────┴──────────────┴─────────┴────────────┘
```

Suppose you want to continue to work on `exp-e6c97` in a separate branch. You
can create a new Git branch by specifying the experiment and giving a new name for it:

```dvc

> $ dvc exp branch exp-e6c97 my-successful-experiment
> Git branch 'my-successful-experiment' has been created from experiment 'exp-e6c97'.
> To switch to the new branch run:
>         git checkout my-successful-experiment
> ```
> 
> Note that DVC doesn't checkout the new branch. You can one or more branches
> from existing experiments, and switch into any one manually like this:
> 
> ```dvc
> $ git checkout my-successful-experiment

Reviewer: my-successful-experiment -> mybranch
````suggestion
$ dvc exp branch exp-e6c97 mybranch
Git branch 'mybranch' has been created from experiment 'exp-e6c97'.
To switch to the new branch run:
        git checkout mybranch
```

Note that DVC doesn't checkout the new branch. You can one or more branches
from existing experiments, and switch into any one manually like this:

```dvc
$ git checkout mybranch
````

Typically, `dvc exp run` leaves the experiment
results in your workspace for convenience. However, you may have run
multiple experiments and wish to go back to a specific one. In this case,
you can restore a previous experiment's results with `dvc exp apply`.
Let's see an example:

```dvc
$ dvc exp show --include-params=my_param
```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Experiment            ┃ Created      ┃     auc ┃ my_param   ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━┩
│ workspace             │ -            │ 0.61314 │ 3          │
│ experiments-tag       │ Oct 19, 2020 │ 0.61314 │ 3          │
│ ├── exp-e6c97         │ Oct 20, 2020 │ 0.69830 │ 2          │
│ └── exp-1df77         │ Oct 22, 2020 │ 0.51676 │ 1          │
└───────────────────────┴──────────────┴─────────┴────────────┘
```

Here, the results found in the workspace in the respective row. When you want to
bring another experiment from the listed experiments, you can do so with:

```dvc
$ dvc exp apply exp-e6c97
Changes for experiment 'exp-e6c97' have been applied...
```

Now, if you list the experiments again with `dvc exp show`, you'll see that the

> workspace contains the experiment with `my_param` of `2` and corresponding `auc` value.

Reviewer: > the workspace contains the experiment with with `my_param` of `2`

- We should reiterate the ID I think (`exp-e6c97`).
- And also, what does it mean that the workspace "contain an experiment"? We should be explicit: both the setup (dependencies, params, etc.) including data and code, and the results (artifacts or ML model in this case? and metrics).

Author: Not all of them are restored from DVC cache, textual content is restored from Git objects. I'll revise this but I think it's better not to mention DVC cache explicitly here.

Author: `dvc exp show` doesn't show an experiment id when it's in the workspace. It just says `workspace.` I didn't put the experiment ID there because of this. `exp-e6c97` becomes "the experiment in the `workspace`" after `dvc exp apply`.

Reviewer: > dvc exp show doesn't show an experiment id when it's in the workspace

Yeah I meant in the text e.g. "you'll see that the workspace contains the results of experiment `exp-e6c97`..."

> Not all of them are restored from DVC cache

OK you can just end the sentence at "restored." Or alternatively just say "contains the results of" like I suggested above and don't go into details.

But I see this whole section was removed? Do you think it's not needed? Up to you then

Author: I didn't remove the section, changed the order of sections. This is done in dd889ca.

Reviewer: Looks good!

> You can use standard Git commands to add, commit and push to a Git repository.
> DVC will take care of the artifacts added to DVC cache, and the rest (params,
> code, text files) will be pushed the Git repository.

Reviewer: Copy edit
```suggestion
You can now use standard Git commands (e.g. `git add/commit/push`)
to version this experiment directly in the <abbr>repository</abbr>.
DVC-tracked data and artifacts are already in the DVC cache, and the rest (params,
code and config files, etc.) can be stored in Git.
```

# Persisting Experiments

> DVC runs experiments outside of the Git stage/commit cycle for quick iteration.

Reviewer: This motivation is not very clear. Let's just use the text from https://dvc.org/doc/user-guide/experiment-management#persistent-experiments probably

> When your experiments are good enough to save or share, you may want to store them persistently as Git commits in your repository.

```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Experiment            ┃ Created      ┃     auc ┃ my_param   ┃
┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━┩
│ workspace             │ -            │ 0.61314 │ 3          │
│ experiments-tag       │ Oct 19, 2020 │ 0.61314 │ 3          │
│ ├── exp-e6c97         │ Oct 20, 2020 │ 0.69830 │ 2          │
│ └── exp-1df77         │ Oct 22, 2020 │ 0.51676 │ 1          │
└───────────────────────┴──────────────┴─────────┴────────────┘
```

Here, the results found in the workspace in the respective row. When you want to
bring another experiment from the listed experiments, you can do so with:

```dvc
$ dvc exp apply exp-e6c97
Changes for experiment 'exp-e6c97' have been applied...
```

Now, if you list the experiments again with `dvc exp show`, you'll see that the
workspace contains the experiment with `my_param` of `2` and corresponding `auc`
value.

You can now use standard Git commands (e.g. `git add/commit/push`) to version
this experiment directly in the <abbr>repository</abbr>. DVC-tracked data and
artifacts are already in the DVC cache, and the rest (params, code and config

> files, etc.) can be stored in Git.

Reviewer: > DVC-tracked data and artifacts are already in the DVC cache

But we should remind the user to `dvc push` in order to share/backup if needed.

Author: I added a sentence without ⚠️ but can move it too.

## Create a Git branch for your experiment

> You may desire to keep the experiments in their separate Git branches. You can

Reviewer: > keep the experiments in their separate Git branches

The user isn't aware that experiments are already branches (and shouldn't know) so we should rephrase this. What's the motivation here?

Reviewer: We cna probably just remove that sentence and open with "You can use `dvc exp branch` **instead**, to create..."

Author: "Instead" in the first sentence seems a bit out of context. "Instead of what?" Deleted the first sentence for now. I'll move this section to the top and rephrase it.

Reviewer: > "Instead" in the first sentence seems a bit out of context. "Instead of what?"

Wasn't the `apply` section before the `branch` section before?

use `dvc exp branch` to create a new branch from the experiment and keep all

> code and artifacts in a new branch.

Reviewer: Copy edit
```suggestion
code and artifacts separate from your current one.
```

```dvc
$ dvc exp show --include-params=my_param
```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━┓

> ┃ Experiment            ┃ Created      ┃     auc ┃ my_param   ┃

Reviewer: Color

Author: Done. Thanks.

┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━┩
│ workspace             │ -            │ 0.61314 │ 3          │

> │ experiments-tag       │ Oct 19, 2020 │ 0.61314 │ 3          │

Reviewer: `experiments-tag`?

Author: I'll rename to it something like `successful-experiment`.

Reviewer: I wouldn't include the name "experiment" in the commit label that contains the experiments. That's confusing. Just `version-x` or something like that maybe?

Author: The tags are related with experiments, not some linear progress. It's like a bucket label, so keeping `experiments` seems more appropriate to me.

│ ├── exp-e6c97         │ Oct 20, 2020 │ 0.69830 │ 2          │
│ └── exp-1df77         │ Oct 22, 2020 │ 0.51676 │ 1          │
└───────────────────────┴──────────────┴─────────┴────────────┘
```

Suppose you want to continue to work on `exp-e6c97` in a separate branch. You

> give a new name to this branch and create a new branch with it.

Reviewer: ```suggestion
can create a new Git branch by specifying the experiment and giving a new name for it:
```

it's name or ID, e.g.:

```dvc
$ dvc exp apply exp-e6c97
Changes for experiment 'exp-e6c97' have been applied...
```

⚠️ Note that, `dvc exp apply` requires you're at the same commit that the
experiment run. Otherwise, it won't apply the experiments due to possible
conflicts.

Now, if you list the experiments again with `dvc exp show`, you'll see that the

> workspace contains the results of `exp-e6c97`.
> 
> You can use standard Git commands (e.g. `git add/commit/push`) to version this
> experiment directly in the <abbr>repository</abbr>. DVC-tracked data and

Reviewer: 💅 
```suggestion
workspace contains the results of `exp-e6c97`.
You can use standard Git commands (e.g. `git add/commit/push`) to version this
experiment directly in the <abbr>repository</abbr> as a regular project version. DVC-tracked data and
```

> Note that DVC doesn't checkout the new branch. You can create as many branches
> from the experiments, and checkout them manually by:
> 
> ```dvc
> $ git checkout my-successful-experiment
> ```

Reviewer: Copy edit
```suggestion
Note that DVC doesn't checkout the new branch. You can one or more branches
from existing experiments, and switch into any one manually with `git checkout`.
```

(No need to teach Git.)

> Note that DVC doesn't checkout the new branch. You can create as many branches
> from the experiments, and checkout them manually by:
> 
> ```dvc
> $ git checkout my-successful-experiment
> ```
> 
> Your workspace now contains the files from the experiment. To update the DVC
> tracked files, you may need to checkout from the cache.
> 
> ```dvc
> $ dvc checkout
> ```

Reviewer: Copy edit
````suggestion
Note that DVC doesn't checkout the new branch. You can one or more branches
from existing experiments, and switch into any one manually like this:

```dvc
$ git checkout my-successful-experiment
$ dvc checkout
```

Your workspace now contains all the files from the experiment.
````

# Persisting Experiments

When you run an experiment with `dvc exp run`, it produces temporary objects in

> the Git repository. These are listed with `dvc exp list` and `dvc exp show`.

Reviewer: We need to put this in the nav sidebar please.

Author: Done! Thanks.

> the Git repository. These are listed with `dvc exp list` and `dvc exp show`.

Reviewer: The part about temporary objects is an implementation detail that we should put in https://dvc.org/doc/user-guide/experiment-management/running-experiments probably. I don't think it's relevant info here.

Author: Removed that too.

> the Git repository. These are listed with `dvc exp list` and `dvc exp show`.

Reviewer: > These are listed with `dvc exp list` and `dvc exp show`.

Not about persisting. Should be in another guide I think (which is why I thought https://github.com/iterative/dvc.org/issues/2844 should come first). We can leave it out of this one unless we're using `exp list/show` for examples here.

Author: Removed. Thanks.

Reviewer: Thank you.

Although the experiments as the temporary, they can be shared with

> `dvc exp push` and retrieved from other Git repositories with `dvc exp pull`. In

Reviewer: `exp push/pull` - also not about persisting. That's already in https://dvc.org/doc/user-guide/experiment-management/sharing-experiments.

Author: Removed them. Thanks.

> Although the experiments as the temporary, they can be shared with
> `dvc exp push` and retrieved from other Git repositories with `dvc exp pull`. In
> this section we'll see how to bring them to the standard Git workflow with
> `dvc exp apply` and `dvc exp branch`.

Reviewer: This one sentence is enough intro for the doc I think:

```suggestion
Let's see how to bring `dvc experiments` to the standard Git workflow with
`dvc exp apply` and `dvc exp branch`.
```

Reviewer: Or maybe add some more motivation as to why you'd want to bring exps to a regular git flow.

Author: I added an initial sentence and introduced the commands.

## Bring Experiment results to your workspace

> If it's not run with `--temp` or `--queue` flags, `dvc exp run` leaves the final
> artifacts in your workspace. This is for convenience. However if there are more
> than one experiment you have run, like
> 
> ```dvc
> $ dvc exp run -S my_param=1
> $ dvc exp run -S my_param=2
> $ dvc exp run -S my_param=3
> ```
> 
> only the final experiment's results are found in the workspace. In this case,
> you may need to bring the previous experiments' results with `dvc exp apply`.

Reviewer: ```suggestion
Typically, `dvc exp run` leaves the experiment
results in your workspace for convenience. However, you may have run
multiple experiments and wish to go back to a specific one, e.g.

```dvc
$ dvc exp run -S my_param=1
$ dvc exp run -S my_param=2
$ dvc exp run -S my_param=3
```

In this case,
you can restore a previous experiment's results with `dvc exp apply`:
```

Reviewer: That's just a copy edit so I'll commit it.

```dvc
$ dvc exp show --include-params=my_param
```

```dvctable
┏━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━┓

> ┃ Experiment            ┃ Created      ┃     auc ┃ my_param   ┃

Reviewer: No colors?

Author: should be fixed now.

┡━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━┩
│ workspace             │ -            │ 0.61314 │ 3          │

> │ experiments-tag       │ Oct 19, 2020 │ 0.61314 │ 3          │

Reviewer: What's `experiments-tag`?

Author: It's the tag we gave to that commit that contains the experiment. Is it confusing? Let me change it then.

Reviewer: Well if it's a real tag in the example repo then OK (unless you want to change it to something more meaningful) but should we at least explain what it is under this image? Maybe as a note (md quote block).

Reviewer: > if it's a real tag in the example repo then OK

Actually we don't need to have the real tags here since it's not a tutorial or even get started. You can change the labels and exp names as needed even if they're imaginary.

Reviewer: This is probably minor. I just think if someone tries to fully understand the table they may be confused by that revision name. I'd suggest something very obvious like keep `HEAD` or do something like `v1.2.3`. Up to you

Author: it's certainly not `HEAD` and `v1.2.3` is not something we'd use in an experimentation setting, there is no linear progress. Changed to `new-experiments` but can replace it if you have a better idea.

│ ├── exp-e6c97         │ Oct 20, 2020 │ 0.69830 │ 2          │
│ └── exp-1df77         │ Oct 22, 2020 │ 0.51676 │ 1          │
└───────────────────────┴──────────────┴─────────┴────────────┘
```

> Here, the results found in the workspace in the respective row. When you want to
> bring another experiment from the listed experiments, you can do so with:

Reviewer: Copy edit (committing)
```suggestion
The results found in the workspace can be found in the respective row. When you want to
bring another experiment to the workspace, you can refer to it using it's name or ID, e.g.:
```

```dvc
$ dvc exp apply exp-e6c97
Changes for experiment 'exp-e6c97' have been applied...
```

Now, if you list the experiments again with `dvc exp show`, you'll see that the

> workspace contains the experiment with `my_param=2` and its `auc` metric.

Reviewer: ```suggestion
workspace contains the experiment with `my_param` of `2` and corresponding `auc` value.
```

Reviewer: (copy edit)
